File: apps/windows/src/PCMobileLink.Windows.Core/PairingShortCode.cs
using System.Security.Cryptography;

namespace PCMobileLink.Windows.Core;

public static class PairingShortCode
{
    private const int CodeLength = 9;
    private const string Alphabet = "23456789ABCDEFGHJKMNPQRSTUVWXYZ";

    public static string Generate()
    {
        Span<char> characters = stackalloc char[CodeLength];
        byte[] randomBytes = RandomNumberGenerator.GetBytes(CodeLength);
        for (int index = 0; index < CodeLength; index++)
        {
            characters[index] = Alphabet[randomBytes[index] % Alphabet.Length];
        }

        return new string(characters);
    }

    public static string Normalize(string value)
    {
        return new string(value.Where(char.IsLetterOrDigit).Select(char.ToUpperInvariant).ToArray());
    }

    public static bool IsValid(string value)
    {
        string normalized = Normalize(value);
        return normalized.Length == CodeLength && normalized.All(character => Alphabet.Contains(character));
    }

    public static string Format(string value)
    {
        string normalized = Normalize(value);
        return normalized.Length == CodeLength
            ? $"{normalized[..3]}-{normalized.Substring(3, 3)}-{normalized.Substring(6, 3)}"
            : value;
    }
}

File: apps/android/app/src/main/java/com/pcmobilelink/nearshare/pairing/PairingShortCode.kt
package com.pcmobilelink.nearshare.pairing

import java.security.SecureRandom

object PairingShortCode {
    private const val CODE_LENGTH = 9
    private const val ALPHABET = "23456789ABCDEFGHJKMNPQRSTUVWXYZ"
    private val random = SecureRandom()

    fun generate(): String {
        val bytes = ByteArray(CODE_LENGTH)
        random.nextBytes(bytes)
        return buildString(CODE_LENGTH) {
            bytes.forEach { byte ->
                val index = (byte.toInt() and 0xFF) % ALPHABET.length
                append(ALPHABET[index])
            }
        }
    }

    fun normalize(value: String): String {
        return value.filter { it.isLetterOrDigit() }.uppercase()
    }

    fun isValid(value: String): Boolean {
        val normalized = normalize(value)
        return normalized.length == CODE_LENGTH && normalized.all { character -> ALPHABET.contains(character) }
    }

    fun format(value: String): String {
        val normalized = normalize(value)
        return if (normalized.length == CODE_LENGTH) {
            "${normalized.substring(0, 3)}-${normalized.substring(3, 6)}-${normalized.substring(6, 9)}"
        } else {
            value
        }
    }
}
